File: cpp/TemplateMethodDemo.cpp
#include <iostream>
#include <string>
using namespace std;

class TvOperator {
	public:
	virtual ~TvOperator() = default;
	virtual void open() = 0;
	virtual void close() = 0;
	void play(){
		open();
		cout<<"playing..."<<endl;
		close();
	}
};

class XiaomiTvOp: public TvOperator {
public:
	void open() override {
		cout<<"Xiaomi TV open"<<endl;
	} 
	void close() override {
		cout<<"Xiaomi TV close"<<endl;
	}

};

class KangjiaTvOp: public TvOperator {
	void open() override {
		cout<<"Kangjia TV open"<<endl;
	} 
	void close() override {
		cout<<"Kangjia TV close"<<endl;
	}
};


int main(int argc, char* argv []) 
{
	TvOperator* tv1 = new XiaomiTvOp;
	tv1->play();
	delete tv1;	

	cout<<endl;

	TvOperator* tv2 = new KangjiaTvOp;
	tv2->play();
	delete tv2;	
}

File: java/behavioral/TemplateMethodDemo.java
//author sbkun
package behavioral;

abstract class TvOperator {
    abstract void open();
    abstract void close();
    void play() {
        open();
        System.out.println("playing...");
        close();
    }
}

class XiaomiTvOp extends TvOperator {
    @Override
    public void open() {
        System.out.println("Xiami TV open");
    }

    @Override
    public void close() {
        System.out.println("Xiaomi TV close");
    }

};

class KangjiaTvop extends TvOperator {
    @Override
    public void open() {
        System.out.println("Kangjia TV open");
    }

    @Override
    public void close() {
        System.out.println("Kangjia TV close");
    }

};

public class TemplateMethodDemo {
    public static void main(String args[]) {
        TvOperator tv1 = new XiaomiTvOp();
        TvOperator tv2 = new KangjiaTvop();
        tv1.play();
        System.out.println("\n");
        tv2.play();
    }
}
